lab5/src/Parser.java

```diff
@@ -0,0 +1,290 @@
+import com.google.common.collect.Table;
+import com.google.common.collect.TreeBasedTable;
+import javafx.util.Pair;
+
+import java.io.IOException;
+import java.util.*;
+
+public class Parser {
+    private Grammar grammar;
```

Comment:
The fields grammar, rules, and productionsNumbered could have also been marked as final.